apps/ai-pipeline/app/pipeline/validation.py

```
@@ -0,0 +1,108 @@
+    """트리 구조 검증"""
+    errors = []
+
+    # 참조되는 노드 ID 수집
+    referenced_ids = {tree.root_node_id}
+    for node in tree.nodes.values():
+        for choice in node.choices:
+            if choice.next_node_id:
+                referenced_ids.add(choice.next_node_id)
+
+    # 1. 고아 노드 검사 (루트가 아닌데 참조되지 않는 노드)
+    for node_id, node in tree.nodes.items():
```

Ruff (0.15.12)

[warning] Loop control variable `node` not used within loop body

(B007)